### DISCUSSION TITLE
Sublime Text syntax plugin

### DISCUSSION THREAD

**Discussion by yaroslavyaroslav** [apple/pkl#300] [Show and tell]
Hey y'all. I've played a bit with the sources that you've provided and was able to port VSCode [syntax](https://github.com/apple/pkl-vscode) to Sublime Text. Here it is: https://github.com/yaroslavyaroslav/pkl-sublimetext

Not a big deal, as I said in readme just replaced hardcoded `JsonRenderer` to `Plist` one, and thanks to you "it just worked".

It would take some time till I'll manage to add this to Package Control repo to make installation straightforward, so I'll just put it here to let the folks know that it's exist.

UPD: I believe there's a plenty room for automating release process by ci/cd somehow, if `JsonRenderer` wasn't hardcoded in the vsc plugin repo.

**Comment by ghyatzo** [comment]
Hi! 
There is already a pkl syntax plugin for sublime in the works! https://packagecontrol.io/packages/Pkl%20%28Pickle%29
I'm not the original author but I have already made some contributions and I am actively working on it.

While waiting on the pacakge author to merge the changes, I am actively working on my fork: https://github.com/ghyatzo/pkl.tmbundle
There are still some minor fixes to be done (at least, when using it with my color-scheme) but other than that it is a decent baseline.
Any testing or help will be much appreciated!

It is based on the same files you used to generate your `.tmLanguage` but rewritten by hand to a `.sublime-syntax` format and extended it with some better scoping and pattern matching.

EDIT: changes have been merged into the pacakge!

**Reply by yaroslavyaroslav** [reply]
Hey, yeah, I've found it in st discord right after I've made this one by myself. 

And I think it worth to focus on a single one instance, the one that is already published, so I'll not pushing this one any further. 

Anyway I saw that the given plugin hand written in ST4 format and have thought that it could worth to implement pkl template for such format at first and than fulfill it with the pkl implementation that just would generates the required yaml after. 

But as for me personally, I decided to move back to this task [st4 template] when the pkl itself gains the promised LSP support.

**Reply by ghyatzo** [reply]
I also contemplated the idea of writing a pkl template for `.sublime-syntax` files since it is "just" YAML. But after toying manually with the format it became clear that there is an extra layer of logic (stack, variables, etc) to it that to properly encode into a pkl render would basically mean write it manually anyway but with one extra layer of abstaction. And ended up just writing it directly.

**Reply by yaroslavyaroslav** [reply]
Isn't such additional layer is the main reason of existing Pkl at all? The one that provides strict typing system, loops and lsp support (there's lack of such for yaml). 

So my idea still is to wait till they implement LSP support and than to drift there from a plain yaml for sake of convenience and the ability to rely on tools in (very) complex configuration files which is the case for st syntax. 

Do I miss something here?

**Reply by ghyatzo** [reply]
[sublime syntax](https://www.sublimetext.com/docs/syntax.html#meta-patterns) is not a plain YAML file full of regexes. It uses the YAML format to define a pseudo language with which you can define contexes and logic on how to manipulate a context stack.
Trying to generate the syntax file with Pkl would mean first wrapping the sublime-syntax logic for manipulating the context stack in pkl and then write basically the same thing you would write manually but with the extra layer of pkl translating it into the correct yaml format.

Hopefully my rambling is clear enough :)
